File: backend/app/schemas/admin_content.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field

SortOrder = Literal["asc", "desc"]
ContentStatus = Literal["active", "deleted", "quarantined", "all"]


class AdminOwnerSnapshot(BaseModel):
    id: Optional[str] = None
    email: Optional[str] = None
    full_name: Optional[str] = None
    role: Optional[str] = None


class AdminReasonRequest(BaseModel):
    reason: str = Field(..., min_length=1, max_length=500)


class AdminDocumentSummary(BaseModel):
    id: str
    original_filename: str
    owner: AdminOwnerSnapshot
    file_type: str
    file_size: int
    uploaded_at: datetime
    processing_status: str
    page_count: Optional[int] = None
    chunk_count: int = 0
    question_count: int = 0
    knowledge_verification_status: Optional[str] = None
    latest_error: Optional[str] = None
    is_quarantined: bool = False
    deleted_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class AdminDocumentDetail(AdminDocumentSummary):
    media_kind: str = "document"
    cloudinary_resource_type: Optional[str] = None
    processing_history: list[dict[str, Any]] = Field(default_factory=list)


class AdminDocumentListResponse(BaseModel):
    items: list[AdminDocumentSummary]
    total: int
    page: int
    page_size: int
    total_pages: int
    generated_at: datetime


class AdminQuestionSummary(BaseModel):
    id: str
    question_set_id: str
    question_index: int
    question_preview: str
    question_type: Optional[str] = None
    difficulty: Optional[str] = None
    subject: Optional[str] = None
    topic: Optional[str] = None
    source_document_id: Optional[str] = None
    source_document_name: Optional[str] = None
    owner: AdminOwnerSnapshot
    citation_status: Optional[str] = None
    hallucination_risk: Optional[str] = None
    moderation_status: str = "draft"
    created_at: datetime
    updated_at: Optional[datetime] = None
    deleted_at: Optional[datetime] = None


class AdminQuestionDetail(AdminQuestionSummary):
    question: str
    options: Optional[dict[str, str]] = None
    correct_answer: str
    explanation: str
    tags: list[str] = Field(default_factory=list)
    bloom_level: Optional[str] = None
    evidence: list[dict[str, Any]] = Field(default_factory=list)


class AdminQuestionListResponse(BaseModel):
    items: list[AdminQuestionSummary]
    total: int
    page: int
    page_size: int
    total_pages: int
    generated_at: datetime


class AdminQuestionUpdateRequest(BaseModel):
    question: Optional[str] = Field(None, min_length=1, max_length=4000)
    options: Optional[dict[str, str]] = None
    correct_answer: Optional[str] = Field(None, min_length=1, max_length=1000)
    explanation: Optional[str] = Field(None, min_length=1, max_length=8000)
    difficulty: Optional[Literal["easy", "medium", "hard"]] = None
    question_type: Optional[Literal["multiple_choice", "true_false", "short_answer"]] = None
    bloom_level: Optional[Literal["remember", "understand", "apply", "analyze"]] = None
    tags: Optional[list[str]] = None
    reason: Optional[str] = Field(default=None, max_length=500)


class AdminQuestionModerationRequest(BaseModel):
    status: Literal["draft", "review_pending", "approved", "published"] = "approved"
    reason: Optional[str] = Field(default=None, max_length=500)


class AdminExamSummary(BaseModel):
    id: str
    name: str
    owner: AdminOwnerSnapshot
    question_count: int
    created_at: datetime
    last_exported_at: Optional[datetime] = None
    status: str
    source_document_id: Optional[str] = None
    source_document_name: Optional[str] = None
    deleted_at: Optional[datetime] = None


class AdminExamListResponse(BaseModel):
    items: list[AdminExamSummary]
    total: int
    page: int
    page_size: int
    total_pages: int
    generated_at: datetime

File: frontend/src/types/adminContent.ts
import type { SortOrder } from './adminUsers';

export type ContentStatus = 'active' | 'deleted' | 'quarantined' | 'all';

export interface AdminOwnerSnapshot {
  id: string | null;
  email: string | null;
  full_name: string | null;
  role: string | null;
}

export interface AdminDocumentSummary {
  id: string;
  original_filename: string;
  owner: AdminOwnerSnapshot;
  file_type: string;
  file_size: number;
  uploaded_at: string;
  processing_status: string;
  page_count: number | null;
  chunk_count: number;
  question_count: number;
  knowledge_verification_status: string | null;
  latest_error: string | null;
  is_quarantined: boolean;
  deleted_at: string | null;
  updated_at: string | null;
}

export interface AdminDocumentDetail extends AdminDocumentSummary {
  media_kind: string;
  cloudinary_resource_type: string | null;
  processing_history: Array<Record<string, unknown>>;
}

export interface AdminDocumentListParams {
  page?: number;
  page_size?: number;
  search?: string;
  user_id?: string;
  file_type?: string;
  processing_status?: string;
  status?: ContentStatus;
  created_from?: string;
  created_to?: string;
  has_error?: boolean;
  knowledge_verification_status?: string;
  sort_by?: string;
  sort_order?: SortOrder;
}

export interface AdminDocumentListResponse {
  items: AdminDocumentSummary[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  generated_at: string;
}

export interface AdminQuestionSummary {
  id: string;
  question_set_id: string;
  question_index: number;
  question_preview: string;
  question_type: string | null;
  difficulty: string | null;
  subject: string | null;
  topic: string | null;
  source_document_id: string | null;
  source_document_name: string | null;
  owner: AdminOwnerSnapshot;
  citation_status: string | null;
  hallucination_risk: string | null;
  moderation_status: string;
  created_at: string;
  updated_at: string | null;
  deleted_at: string | null;
}

export interface AdminQuestionDetail extends AdminQuestionSummary {
  question: string;
  options: Record<string, string> | null;
  correct_answer: string;
  explanation: string;
  tags: string[];
  bloom_level: string | null;
  evidence: Array<Record<string, unknown>>;
}

export interface AdminQuestionListParams {
  page?: number;
  page_size?: number;
  search?: string;
  user_id?: string;
  document_id?: string;
  question_type?: string;
  difficulty?: string;
  moderation_status?: string;
  status?: ContentStatus;
  created_from?: string;
  created_to?: string;
  sort_order?: SortOrder;
}

export interface AdminQuestionListResponse {
  items: AdminQuestionSummary[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  generated_at: string;
}

export interface AdminQuestionUpdatePayload {
  question?: string;
  options?: Record<string, string>;
  correct_answer?: string;
  explanation?: string;
  difficulty?: 'easy' | 'medium' | 'hard';
  question_type?: 'multiple_choice' | 'true_false' | 'short_answer';
  bloom_level?: 'remember' | 'understand' | 'apply' | 'analyze';
  tags?: string[];
  reason?: string;
}

export interface AdminExamSummary {
  id: string;
  name: string;
  owner: AdminOwnerSnapshot;
  question_count: number;
  created_at: string;
  last_exported_at: string | null;
  status: string;
  source_document_id: string | null;
  source_document_name: string | null;
  deleted_at: string | null;
}

export interface AdminExamListParams {
  page?: number;
  page_size?: number;
  search?: string;
  user_id?: string;
  status?: ContentStatus;
  created_from?: string;
  created_to?: string;
  sort_by?: string;
  sort_order?: SortOrder;
}

export interface AdminExamListResponse {
  items: AdminExamSummary[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  generated_at: string;
}
